mica-web-model/src/main/java/org/obiba/mica/web/model/DatasetDtos.java

```diff
@@ -152,24 +150,31 @@
       }
     }
 
-    Optional.ofNullable(dataset.getStudyId())
-      .ifPresent(studyId -> Optional.ofNullable(dataset.getPopulationId())
-        .ifPresent(populationId -> {
-          Mica.HarmonizationDatasetDto.HarmonizationStudyReferenceDto.Builder sRefBuilder
-            = Mica.HarmonizationDatasetDto.HarmonizationStudyReferenceDto.newBuilder();
-
-          if(asDraft) {
-            if(subjectAclService.isPermitted("/draft/harmonization-study", "VIEW", studyId)) {
-              sRefBuilder.setStudyId(studyId);
-              sRefBuilder.setPopulationId(populationId);
-            }
-          } else if(publishedStudyService.findById(studyId) != null) {
-            sRefBuilder.setStudyId(studyId);
-            sRefBuilder.setPopulationId(populationId);
-          }
-          sRefBuilder.setStudySummary(studySummaryDtos.asHarmoStudyDto(studyId));
-          hbuilder.setStudyReference(sRefBuilder);
-        }));
+    if (dataset.getHarmonizationLink() != null) {
+      HarmonizationTable harmonizationLink = dataset.getHarmonizationLink();
+      Mica.DatasetDto.HarmonizationTableDto.Builder harmonizationLinkBuilder = Mica.DatasetDto.HarmonizationTableDto.newBuilder();
+
+      if (!Strings.isNullOrEmpty(harmonizationLink.getProject()))
+        harmonizationLinkBuilder.setProject(harmonizationLink.getProject());
+
+      if (!Strings.isNullOrEmpty(harmonizationLink.getTable()))
+        harmonizationLinkBuilder.setTable(harmonizationLink.getTable());
+
+      String studyId = harmonizationLink.getHarmonizationStudyId();
+
+      if (asDraft) {
+        if(studyId != null && subjectAclService.isPermitted("/draft/harmonization-study", "VIEW", studyId)) {
+          harmonizationLinkBuilder.setStudyId(harmonizationLink.getHarmonizationStudyId());
+          harmonizationLinkBuilder.setPopulationId(harmonizationLink.getPopulationId());
+        }
+      } else if (studyId != null && publishedStudyService.findById(studyId) != null) {
+        harmonizationLinkBuilder.setStudyId(harmonizationLink.getHarmonizationStudyId());
+        harmonizationLinkBuilder.setPopulationId(harmonizationLink.getPopulationId());
+      }
+
+      if (studyId != null) harmonizationLinkBuilder.setStudySummary(studySummaryDtos.asHarmoStudyDto(studyId));
+      hbuilder.setHarmonizationLink(harmonizationLinkBuilder);
+    }
```

Comment:
Too big method, is it possible to make smaller private methods (Self-documenting code)